tests/save_load.test.js

```diff
@@ -0,0 +1,226 @@
+    });
+
+    const promiseRemovedListener = new Promise((resolve) => {
+      removeListener.mockImplementation(() => resolve());
+    });
+
+    let listener;
+    addListener.mockImplementation((callback) => {
+      listener = callback;
+    });
+
+    download.mockImplementation(() => {
+      const id = 123;
+      setTimeout(() => {
+        listener({
+          state: { current: 'in_progress' },
+          id,
+        });
+        listener({
+          state: { current: 'complete' },
+          id,
+        });
+      }, 0);
+      return Promise.resolve(id);
+    });
+
+    const log = [{ prop1: 'log1' }];
+    await save.saveAsJSON(log);
+
+    Promise.all([promiseRevokedURL, promiseRemovedListener]);
+
+    const blob = downloadFileFn.mock.calls[0][0];
+    const blobContent = await load.readFile(blob);
+
+    expect(blobContent).toBe(JSON.stringify(log));
+
+    expect(downloadFileFn).toHaveBeenCalledTimes(1);
+    expect(download).toHaveBeenCalledTimes(1);
+    expect(download).toHaveBeenCalledWith({
+      url: 'fake-blob-url',
+      filename: 'activitylogs.json',
+    });
+
+    expect(addListener).toHaveBeenCalledTimes(1);
+    expect(addListener).toHaveBeenCalledWith(listener);
+
+    expect(removeListener).toHaveBeenCalledTimes(1);
+    expect(removeListener).toHaveBeenCalledWith(listener);
+
+    expect(revokeObjectURL).toHaveBeenCalledTimes(1);
+    expect(revokeObjectURL).toHaveBeenCalledWith('fake-blob-url');
+
+    downloadFileFn.mockClear();
+  });
+
+  test('Save saveAsJSON should return error message if error is encountered', async () => {
+    const downloadFileFn = jest.spyOn(save, 'downloadFile');
+
+    const createObjectURL = jest.fn();
+    const revokeObjectURL = jest.fn();
+    window.URL = { createObjectURL, revokeObjectURL };
+
+    const download = jest.fn();
+    const addListener = jest.fn();
+    const removeListener = jest.fn();
+    window.browser = {
+      downloads: {
+        download,
+        onChanged: { addListener, removeListener },
+      },
+    };
+
+    createObjectURL.mockReturnValue('fake-blob-url');
+
+    const promiseRevokedURL = new Promise((resolve) => {
+      revokeObjectURL.mockImplementation(() => resolve());
+    });
+
+    let listener;
+    addListener.mockImplementation((callback) => {
+      listener = callback;
+    });
+
+    download.mockImplementation(() => {
+      const id = 123;
+      setTimeout(() => {
+        listener({
+          state: { current: 'interrupted' },
+          error: 'save-error',
+        });
+      }, 0);
+      return Promise.resolve(id);
+    });
+
+    const promiseRemovedListener = new Promise((resolve) => {
+      removeListener.mockImplementation(() => resolve());
+    });
+
+    const log = [{ prop1: 'log1' }];
+
+    const saveAsJson = save.saveAsJSON(log);
+    await expect(saveAsJson).rejects.toThrow('save-error');
+
+    Promise.all([promiseRevokedURL, promiseRemovedListener]);
```

Comment:
This has not been addressed yet: https://github.com/mozilla/extension-activity-monitor/pull/2#discussion_r440122183

I think that you misunderstood me. By "don't use `await`", I didn't mean to literally remove `await` (using `Promise.all` without using/awaiting the resulting promise is not useful). I meant that after returning from `readFile`, that the cleanup (revokeObjectURL and removeListener) should have happened already. So you shouldn't use `promiseRevokedURL` (etc., remove it), but verify that the spies have been called as expected (already done below).

Reply:
Done now :-)